### PR TITLE
T13546

### PR DESCRIPTION
**Pull request by smspillaz**
Implement flipping back to the main window. The implementation for this is awful, but we can at least test using it.

### CODE REVIEW

File: js/ui/windowManager.js [hunk 1]
```diff
@@ -922,17 +922,45 @@ const WindowManager = new Lang.Class({
...
             }
         }
 
-        this._pairedWindows.push([src, dst]);
+        if (!amendedPair) {
+            this._pairedWindows.push([src, dst]);
+        }
+
+        /* The way this works is not ideal at the moment since it
+         * always flips left (can't know what direction to flip in */
```
Comment (cosimoc) [inline js/ui/windowManager.js]: Missing closing paren in comment

File: js/ui/windowManager.js [hunk 2]
```diff
@@ -748,6 +748,8 @@ const WindowManager = new Lang.Class({
                                                        '/com/endlessm/Sylvester/Service');
         this._sylvesterListener.connectSignal('RotateBetweenPidWindows',
                                               Lang.bind(this, this._handleRotateBetweenPidWindows));
+        this._sylvesterListener.connectSignal('RevertXidRotation',
```
Comment (smspillaz) [inline js/ui/windowManager.js]: Ah yes, I completely forgot about pushing that one. Here you go https://github.com/endlessm/sylvester/pull/5

File: src/shell-global.c [hunk 3]
```diff
@@ -1689,3 +1689,10 @@ shell_global_get_session_mode (ShellGlobal *global)
 
   return global->session_mode;
 }
+
+gboolean
+shell_global_window_matches_xid (ShellGlobal *global,
+                                 MetaWindow  *window,
+                                 guint32      xid) {
```
Comment (smspillaz) [inline src/shell-global.c]: Rebased the original commit to keep it upstreamable (if necessary)

File: src/shell-global.c [hunk 4]
```diff
@@ -1689,3 +1689,10 @@ shell_global_get_session_mode (ShellGlobal *global)
...
+    return meta_window_get_xwindow(window) == xid;
```
Comment (smspillaz) [inline src/shell-global.c]: Rebased the original commit to keep it upstreamable (if necessary)

File: js/ui/windowManager.js [hunk 5]
```diff
@@ -807,37 +900,10 @@ const WindowManager = new Lang.Class({
...
-            let dst_geometry = animationSpec.src.rect;
-            animationSpec.dst.window.get_meta_window().move_resize_frame(false,
-                                                                         dst_geometry.x,
-                                                                         dst_geometry.y,
-                                                                         dst_geometry.width,
-                                                                         dst_geometry.height);
-
-            this._rotateInActors.push(animationSpec.dst.window);
-            this._rotateOutActors.push(animationSpec.src.window);
+            let commenceAnimation = prepareWindowsForRotation(animationSpec.src.window,
```
Comment (smspillaz) [inline js/ui/windowManager.js]: Agreed. changed in fea1cf

File: js/ui/windowManager.js [hunk 6]
```diff
@@ -585,6 +585,99 @@ const TilePreview = new Lang.Class({
...
+        /* Tween both windows in a rotation animation at the same time
+         * with backface culling enabled on both. This will allow for
+         * a smooth transition. */
+        Tweener.addTween(src, {
+            rotation_angle_y: direction == RotationDirection.RIGHT ? 180 : -180,
+            time: WINDOW_ANIMATION_TIME * 4,
+            transition: 'easeOutQuad',
+            onComplete: function() {
+                if (srcDone) {
+                    return srcDone(src);
```
Comment (smspillaz) [inline js/ui/windowManager.js]: Fixed in ad1438

File: js/ui/windowManager.js [hunk 7]
```diff
@@ -876,6 +877,60 @@ const WindowManager = new Lang.Class({
...
+         * destination window.
+         *
+         * We use a naive loop here since we're detecting where this
+         * is the case and then breaking out after adding then new
+         * entry. */
+        for (let i = 0; i < this._pairedWindows.length; ++i) {
+            let [pairSrc, pairDst] = this._pairedWindows[i];
+            if (pairDst == src) {
+                this._pairedWindows.splice(i, 1);
+                log("Spliced and inserted cross-cutting pair");
```
Comment (smspillaz) [inline js/ui/windowManager.js]: Hmm, I can't find this in my working copy.

File: js/ui/windowManager.js (hunk 1 above)
Comment (smspillaz) [inline js/ui/windowManager.js]: Fixed

File: js/ui/windowManager.js [hunk 8]
```diff
@@ -780,6 +782,9 @@ const WindowManager = new Lang.Class({
         this._updateReadyRotateAnimationsWith(srcActorInfo.window);
     },
 
+    _handleRotateWindowBack: function(proxy, sender, [xid]) {
```
Comment (smspillaz) [inline js/ui/windowManager.js]: Yes - rebased the original commit and adjusted the message too

File: js/ui/windowManager.js [hunk 9]
```diff
@@ -585,6 +585,99 @@ const TilePreview = new Lang.Class({
     }
 });
 
+/**
+ * The direction a window should rotate in
+ */
+const RotationDirection = {
```
Comment (smspillaz) [inline js/ui/windowManager.js]: 8e41fc - I went with `GtkDirectionType` because it was the same number of characters :)

**Comment by smspillaz** [comment]
Okay, comments should be addressed. I didn't get a chance to re-test this, but since the changes were relatively trivial I think nothing is broken. I'll test again later today.

**Comment by smspillaz** [comment]
@erikos Not sure if we wanted to merge this for now?

**Comment by magcius** [comment]
> I wish there was a better way to group windows this way, but I am pretty sure that X does not really have this feature...

X11 / mutter expert here -- what kind of "group" do you want?

**Comment by smspillaz** [comment]
For the sake of avoiding confusion between this and a similar branch still in development by @erikos , I am going to close this one.

**Pull request closed without merging**